clients/src/main/java/org/apache/kafka/clients/producer/TransactionState.java

```diff
@@ -29,9 +58,123 @@
  * A class which maintains state for transactions. Also keeps the state necessary to ensure idempotent production.
  */
 public class TransactionState {
+    private static final Logger log = LoggerFactory.getLogger(TransactionState.class);
+
+    private static final int NO_INFLIGHT_REQUEST_CORRELATION_ID = -1;
+
     private volatile PidAndEpoch pidAndEpoch;
     private final Map<TopicPartition, Integer> sequenceNumbers;
-    private final Time time;
+    private final String transactionalId;
+    private final int transactionTimeoutMs;
+    private final PriorityQueue<TransactionalRequest> pendingTransactionalRequests;
+    private final Set<TopicPartition> newPartitionsToBeAddedToTransaction;
+    private final Set<TopicPartition> pendingPartitionsToBeAddedToTransaction;
+    private final Set<TopicPartition> partitionsInTransaction;
+    private final Map<TopicPartition, TxnOffsetCommitRequest.CommittedOffset> pendingTxnOffsetCommits;
+    private int inFlightRequestCorrelationId = NO_INFLIGHT_REQUEST_CORRELATION_ID;
+
+    private Node transactionCoordinator;
+    private Node consumerGroupCoordinator;
+    private volatile boolean isInTransaction = false;
```

Comment:
Rather than having all these variables to keep track of the state, could we create an enumeration with all of the valid states (sort of like what we do on the server side). That will make it easier to ensure that state transitions are correct.